### PR TITLE
[Merged by Bors] - refactored blockid to be 20 bytes

### PR DESCRIPTION
**Pull request by antonlerner**
## Motivation
fixes issue #1568 

## Changes
added hash20 type and converted all blockids to hash20, also hacked the sync to take 20 byte addresses

## Test Plan
Passes UT, added tests for converting from hash20 to hash32 and back

### CODE REVIEW

File: common/types/hashes.go [hunk 1]
```diff
@@ -23,6 +24,70 @@ type Hash12 [Hash12Length]byte
 // Hash represents the 32 byte Keccak256 hash of arbitrary data.
 type Hash32 [Hash32Length]byte
 
+// Hash represents the 20 byte Keccak256 hash of arbitrary data.
+type Hash20 [Hash20Length]byte
+
+// HexToHash sets byte representation of s to hash.
+// If b is larger than len(h), b will be cropped from the left.
+//func HexToHash20(s string) Hash20 { return BytesToHash(util.FromHex(s)) }
```
Comment (y0sher) [inline common/types/hashes.go:32]: remove comment ?

File: common/types/hashes.go [hunk 2]
```diff
@@ -177,6 +242,12 @@ func (h *Hash32) SetBytes(b []byte) {
 	copy(h[32-len(b):], b)
 }
 
+func (h Hash32) ToHash20() Hash20 {
```
Comment (y0sher) [inline common/types/hashes.go:245]: maybe we could have copyyohash20 and tohash20 where the latter doesn't copy but use the same array. of course only if we have a use for it.

File: common/types/block.go [hunk 3]
```diff
@@ -86,7 +86,7 @@ type MiniBlock struct {
 }
 
 func (id BlockID) AsHash32() Hash32 {
```
Comment (almogdepaz) [inline common/types/block.go:88]: are we going to change tx ids and atx ids as well ? if so we really don't need to keep AsHash32 only for sync proposes

Comment (antonlerner) [reply common/types/block.go:88]: No, I don't think we should change everything to be 20bytes hash

**Comment by antonlerner** [comment]
bors try

File: common/types/hashes.go (hunk 2 above)
Comment (antonlerner) [reply common/types/hashes.go:245]: this is the safest way to do it and not risk leaking bytes IMO. if we need better performance we could always change it to be non copy

File: common/types/block.go (hunk 3 above)
Comment (barakshani) [reply common/types/block.go:88]: we should decrease them as well (maybe not in this pr), maybe even smaller than 20, as iddo wants

Comment (gavraz) [reply common/types/block.go]: We should change them also. Another PR may address the on-wire vs real-size.
For now, we want to have significantly smaller IDs across the system.

File: common/types/hashes.go (hunk 2 above)
Comment (gavraz) [reply common/types/hashes.go:245]: We should prefer copy over referencing such data.
As I have said before, Hash* should behave like an int (copy) wherever possible.

File: common/types/block.go (hunk 3 above)
Comment (gavraz) [reply common/types/block.go:88]: IMO, we should change tx and atx IDs to Hash20 now.
Later, on another PR, we should distinguish between wire-size and internal-size.

**Review by gavraz — CHANGES_REQUESTED** [review]
In addition to comments, make sure you run all the tests.

File: common/types/block.go (hunk 3 above)
Comment (antonlerner) [reply common/types/block.go:88]: why? it's already implemented this way

Comment (gavraz) [reply common/types/block.go:88]: Right, I thought we changed it too but it was already like that.

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors merge

**Comment by antonlerner** [comment]
bors try

**Review by gavraz — APPROVED** [review]

**Comment by antonlerner** [comment]
bors try

**Comment by antonlerner** [comment]
bors merge

**Comment by antonlerner** [comment]
bors merge

**Pull request closed without merging**